cubi_tk/irods/check.py

```diff
@@ -1,73 +1,100 @@
-"""``cubi-tk irods check``: Check target iRods collection (all md5 files? metadata md5 consistent? enough replicas?)."""
+"""``cubi-tk irods check``: Check target iRODS collection (all md5 files? metadata md5 consistent? enough replicas?)."""
 
-import os
-import sys
 import argparse
+import json
+import os
 import re
+import tqdm
 import typing
-import uuid
-from multiprocessing.pool import ThreadPool
-from subprocess import check_output, SubprocessError
-from retrying import retry
 
+from irods.collection import iRODSCollection
+from irods.data_object import iRODSDataObject
+from irods.session import iRODSSession
 from logzero import logger
-import tqdm
-
+from multiprocessing.pool import ThreadPool
 
 MIN_NUM_REPLICAS = 2
-NUM_PARALLEL_TESTS = 8
+NUM_PARALLEL_TESTS = 4
+MD5_RE = re.compile(r"[0-9a-fA-F]{32}")
 
 
 class IrodsCheckCommand:
-    """Implementation of irods check command."""
+    """Implementation of iRDOS check command."""
 
     command_name = "check"
 
     def __init__(self, args):
         #: Command line arguments.
         self.args = args
 
+        #: Path to iRODS environment file
+        self.irods_env_path = os.path.join(
+            os.path.expanduser("~"), ".irods", "irods_environment.json"
+        )
+
+        #: iRODS sessions for parallel execution
+        self.irods_sessions = []
+
+        #: iRODS environment
+        self.irods_env = None
+
+    def __del__(self):
```

Comment:
This looks like it's prone to concurrency/inderministic issues. Can we switch this to use Python context managers that ensure that the sessions are closed / cleaned more deterministically?

Reply:
I guess if I refactor the whole thing into such that we open new connections for each operation with a context manager and then do a cleanup, that could work. It will probably be slower though. Not sure what the issue here is.

Reply:
Something like this could wokr

```
from contextlib import contextmanager

class IrodsCheckCommand:

    # ...

    @contextmanager
    def _irods_sessions(self, first, count):
        irods_sessions = [first]
        irods_sessions += [self.connect() for _ in range(count - 1)]
        try:
            yield irods_sessions
        finally:
            for irods in irods_sessions:
                irods.cleanup()


    # ...

        # ...
        with self._irods_sessions(session, s_count) as irods_sessions:
            pass  # ...
```

Reply:
Alright, will refactor.

Reply:
Done.